# Fehler

Ein [Sumor Cloud](https://sumor.cloud) Tool.  
[Weitere Dokumentation](https://sumor.cloud/error) Eine Fehlerbehandlungs-Bibliothek mit Unterstützung für mehrere Sprachen, vordefinierter Fehlercode und Datenübergabe.

[![CI](https://github.com/sumor-cloud/error/actions/workflows/ci.yml/badge.svg)](https://github.com/sumor-cloud/error/actions/workflows/ci.yml)
[![Test](https://github.com/sumor-cloud/error/actions/workflows/ut.yml/badge.svg)](https://github.com/sumor-cloud/error/actions/workflows/ut.yml)
[![Abdeckung](https://github.com/sumor-cloud/error/actions/workflows/coverage.yml/badge.svg)](https://github.com/sumor-cloud/error/actions/workflows/coverage.yml)
[![Audit](https://github.com/sumor-cloud/error/actions/workflows/audit.yml/badge.svg)](https://github.com/sumor-cloud/error/actions/workflows/audit.yml)

## Installation

```bash
npm i @sumor/error --save
```

## Voraussetzungen

### Node.JS-Version

Erfordert Node.JS-Version 18.x oder höher

### Erfordert Node.JS ES-Modul

Da dieses Paket in ES-Modul geschrieben ist,
ändern Sie bitte den folgenden Code in Ihrer `package.json`-Datei:

```json
{
  "type": "module"
}
```

## Verwendung

### Fehler definieren

```js
import defineError from '@sumor/error'

const MyError = defineError({
  code: {
    USER_NOT_FOUND: 'Benutzer nicht gefunden',
    USER_EXISTED: 'Benutzer {name} existiert'
  }
})

throw new MyError('USER_NOT_FOUND')
// Ausgabe: Error: Benutzer nicht gefunden

throw new MyError('USER_EXISTED', { name: 'Alice' })
// Ausgabe: Error: Benutzer Alice existiert
```

### Mehrsprachig

```js
import defineError from '@sumor/error'

const MyError = defineError({
  Sprache: 'en', // Standardsprache
  code: {
    USER_NOT_FOUND: 'Benutzer nicht gefunden',
    USER_EXISTED: 'Benutzer {name} existiert'
  },
  i18n: {
    zh: {
      USER_NOT_FOUND: 'Benutzer nicht gefunden',
      USER_EXISTED: 'Benutzer {name} existiert'
    }
  }
})

const error = new MyError('USER_EXISTED', { name: 'Alice' })
error.language = 'en' // Fehler Sprache ändern
console.log(error)
// Ausgabe: Error: Benutzer Alice existiert

error.language = 'zh' // Fehler Sprache ändern
console.log(error)
// Ausgabe: Error: Benutzer Alice existiert
```

### Fehler in JSON konvertieren

```js
import defineError from '@sumor/error'

const MyError = defineError({
  code: {
    USER_NOT_FOUND: 'Benutzer nicht gefunden',
    USER_EXISTED: 'Benutzer {name} existiert'
  }
})

const error = new MyError('USER_EXISTED', { name: 'Alice' })
console.log(error.json())
// Ausgabe: {"code":"USER_EXISTED","message":"Benutzer Alice existiert"}
```

### Unterliegender Fehler

```js
import defineError from '@sumor/error'

const MyError = defineError({
  code: {
    FIELD_VERIFY_FAILED: 'Feldüberprüfung fehlgeschlagen',
    FIELD_CANNOT_EMPTY: 'Feld {name} darf nicht leer sein',
    FIELD_TOO_LONG: 'Feld {name} ist zu lang'
  },
  i18n: {
    zh: {
      FIELD_VERIFY_FAILED: 'Feldüberprüfung fehlgeschlagen',
      FIELD_CANNOT_EMPTY: 'Feld {name} darf nicht leer sein',
      FIELD_TOO_LONG: 'Feld {name} ist zu lang'
    }
  }
})

const error = new MyError('FIELD_VERIFY_FAILED', {}, [
  new MyError('FIELD_CANNOT_EMPTY', { name: 'benutzername' }),
  new MyError('FIELD_TOO_LONG', { name: 'passwort' })
])

console.log(error.json())
/* 
Ausgabe: 
{
  "code":"FIELD_VERIFY_FAILED",
  "message":"Feldüberprüfung fehlgeschlagen",
  "errors":[
    {
      "code":"FIELD_CANNOT_EMPTY",
      "message":"Feld benutzername darf nicht leer sein"
    },{
      "code":"FIELD_TOO_LONG",
      "message":"Feld passwort ist zu lang"
    }
  ]
}
*/

error.language = 'zh'
console.log(error.json())
/*
Ausgabe:
{
  "code":"FIELD_VERIFY_FAILED",
  "message":"Feldüberprüfung fehlgeschlagen",
  "errors":[
    {
      "code":"FIELD_CANNOT_EMPTY",
      "message":"Feld benutzername darf nicht leer sein"
    },{
      "code":"FIELD_TOO_LONG",
      "message":"Feld passwort ist zu lang"
    }
  ]
}
*/
```

### Standardfehler kombinieren

```js
import defineError from '@sumor/error'

const MyError = defineError({
  code: {
    FIELD_VERIFY_FAILED: 'Feldüberprüfung fehlgeschlagen',
    FIELD_CANNOT_EMPTY: 'Feld {name} darf nicht leer sein',
    FIELD_TOO_LONG: 'Feld {name} ist zu lang'
  }
})

const error = new MyError('FIELD_VERIFY_FAILED', {}, [
  new MyError('FIELD_CANNOT_EMPTY', { name: 'benutzername' }),
  new MyError('FIELD_TOO_LONG', { name: 'passwort' }),
  new Error('Unbekannter Fehler')
])

console.log(error.json())
/*
Ausgabe:
{
  "code":"FIELD_VERIFY_FAILED",
  "message":"Feldüberprüfung fehlgeschlagen",
  "errors":[
    {
      "code":"FIELD_CANNOT_EMPTY",
      "message":"Feld benutzername darf nicht leer sein"
    },{
      "code":"FIELD_TOO_LONG",
      "message":"Feld passwort ist zu lang"
    },{
      "code":"UNKNOWN_ERROR",
      "message":"Unbekannter Fehler"
    }
  ]
}
 */
```
